# Accident-level aggregation

This notebook performs data aggregation at the accident level (`accident_number`) because `clean_df` contains multiple rows per accident due to multiple places being recorded.
By aggregating, we can have one accident per row.

In [71]:
import pandas as pd
import numpy as np

In [72]:
clean_df = pd.read_csv('../data/clean_df.csv')

In [73]:
print(f'All count of `accident_number`   : {len(clean_df.accident_number)}')
print(f'Unique count of `accident_number`: {len(clean_df.accident_number.unique())}')

All count of `accident_number`   : 216081
Unique count of `accident_number`: 204145


📝 In the end the length of the df has to be **204145**

---

## The function

- for each categorical column --> taking its mode, and use the first value where the mode cannot be decided
- for numerical columns --> taking the max value 

In [74]:
def mode_or_first(x):
    """
    It returns the mode of a pandas Series.
    If there is a tie, it returns the first value in the Series.
    """
    modes = x.mode()
    if len(modes) == 1:
        return modes.iloc[0]
    else:
        # Where the mode is tie, return the first value
        return x.iloc[0]

In [75]:
%%time

agg_dict = {
    "date": mode_or_first,
    "day_of_week": mode_or_first,
    "hour": mode_or_first,
    "road_category": mode_or_first,
    "road_layout": mode_or_first,
    "num_lanes": "max",
    "reserved_lane": mode_or_first,
    "road_profile": mode_or_first,
    "road_shape": mode_or_first,
    "surface_condition": mode_or_first,
    "infrastructure": mode_or_first,
    "road_location": mode_or_first,
    "speed_limit": "max",
    "injury_severity": "max",
    "users_involved": "max",
    "light_conditions": mode_or_first,
    "department": mode_or_first,
    "urban_area": mode_or_first,
    "intersection_type": mode_or_first,
    "weather": mode_or_first,
    "collision_type": mode_or_first,
    "latitude": mode_or_first,
    "longitude": mode_or_first,
    "region": mode_or_first
}

agg_df = clean_df.groupby("accident_number").agg(agg_dict).reset_index()

CPU times: user 1min 56s, sys: 1.1 s, total: 1min 57s
Wall time: 1min 56s


## Final check

In [76]:
print(f'All count of `accident_number`   : {len(agg_df.accident_number)}')
print(f'Unique count of `accident_number`: {len(agg_df.accident_number.unique())}')

assert len(agg_df) == len(agg_df.accident_number.unique()), "There are duplicate accident_number"

All count of `accident_number`   : 204145
Unique count of `accident_number`: 204145


## Export agg_df to csv

In [77]:
agg_df.to_csv('../data/agg_df.csv', index=False)